File: Python/Arrays/2_validate_subsequence.py
# O(n) time | O(1) space
def validateSubsequence(array, sequence):
  arrIdx = 0
  seqIdx = 0
  while arrIdx < len(array) and seqIdx < len(sequence):
    if array[arrIdx] == sequence[seqIdx]:
      seqIdx += 1
    arrIdx += 1 # regardless of whether or not we find a match, we want to move the array pointer forward
  return seqIdx == len(sequence)

# O(n) time | O(1) space
def validateSubsequence2(array, sequence):
  seqIdx = 0
  for value in array:
    # if we can't find a value in the sequence, then it's not a valid subsequence
    if seqIdx == len(sequence):
      # return True
      break
    if sequence[seqIdx] == value:
      seqIdx += 1
  return seqIdx == len(sequence)

File: Go/Arrays/2_validate_subsequence.go
package main

// O(n) Time | O(1) Space
func IsValidSubsequence(array []int, sequence []int) bool {
	arrIdx := 0
	seqIdx := 0
	for arrIdx < len(array) && seqIdx < len(sequence) {
		if array[arrIdx] == sequence[seqIdx] {
			seqIdx += 1
		}
		arrIdx += 1
	}
	return seqIdx == len(sequence)
}

func IsValidSubsequence2(array []int, sequence []int) bool {
	seqIdx := 0
	for _, value := range array {
		if seqIdx == len(sequence) {
			break
		}
		if value == sequence[seqIdx] {
			seqIdx += 1
		}
	}
	return seqIdx == len(sequence)
}
